File: beginners-docs/C/samples/combobox.py
from gi.repository import Gtk
import sys

distros = [["Select distribution"], ["Fedora"], ["Mint"], ["Suse"]]

class MyWindow(Gtk.ApplicationWindow):
    def __init__(self, app):
        Gtk.Window.__init__(self, title="Welcome to GNOME", application=app)
        self.set_default_size(200, -1)
        self.set_border_width(10)

        # the data in the model, of type string
        listmodel = Gtk.ListStore(str)
        # append the data in the model
        for i in range(len(distros)):
            listmodel.append(distros[i])

        # a combobox to see the data stored in the model
        combobox = Gtk.ComboBox(model=listmodel)

        # a cellrenderer to render the text
        cell = Gtk.CellRendererText()

        # pack the cell into the beginning of the combobox, allocating
        # no more space than needed
        combobox.pack_start(cell, False)
        # associate a property ("text") of the cellrenderer (cell) to a column (column 0)
        # in the model used by the combobox
        combobox.add_attribute(cell, "text", 0)

        # the first row is the active one by default at the beginning
        combobox.set_active(0)

        # connect the signal emitted when a row is selected to the callback function
        combobox.connect("changed", self.on_changed)

        # add the combobox to the window
        self.add(combobox)

    def on_changed(self, combo):
        # if the row selected is not the first one, write its value on the terminal
        if combo.get_active() != 0:
            print "You chose " + str(distros[combo.get_active()][0]) +"."
        return True


class MyApplication(Gtk.Application):
    def __init__(self):
        Gtk.Application.__init__(self)

    def do_activate(self):
        win = MyWindow(self)
        win.show_all()

    def do_startup(self):
        Gtk.Application.do_startup(self)

app = MyApplication()
exit_status = app.run(sys.argv)
sys.exit(exit_status)

File: platform-demos/C/samples/combobox.c
#include <gtk/gtk.h>



/* This is the callback function. It is a handler function which reacts to the
 * signal. In this case, if the row selected is not the first one of the
 * ComboBox, we write its value in the terminal for the user.
 */
static void
on_changed (GtkComboBox *widget,
            gpointer   user_data)
{
  GtkComboBox *combo_box = widget;

  /* This will return the index of the currently active
   * item within the ComboBox.
   */
  gint value = gtk_combo_box_get_active (combo_box);

  /* Equate the index to the list of the choices you put up and
   * print what happened to the user.
   */
  switch (value) {
     case 1:
       g_print ("You chose Fedora.\n");
       break;
     case 2:
       g_print ("You chose Mint.\n");
       break;
     case 3:
       g_print ("You chose Suse.\n");
       break;
     default:
       /* Nothing else will happen in the terminal otherwise */
       break;
   }
}



static void
activate (GtkApplication *app,
          gpointer        user_data)
{
  GtkWidget *window;
  GtkWidget *combo_box;
  GtkWidget *view;

  /* Create a window with a title, border width, and a default size. Setting the
   * size to -1 means to use the "natural" default size.
   * (the size request of the window)
   */
  window = gtk_application_window_new (app);
  gtk_window_set_title (GTK_WINDOW (window), "Welcome to GNOME");
  gtk_window_set_default_size (GTK_WINDOW (window), 200, -1);
  gtk_container_set_border_width (GTK_CONTAINER (window), 10);

  /* Create the combo box and append your string values to it. Note that the
   * values in our combo box are to be this:
   * @ Position 0 - "Select distribution"
   * @ Position 1 - "Fedora"
   * @ Position 2 - "Mint"
   * @ Position 3 - "Suse"
   * (For a more advanced approach, try putting these into a string array)
   * Furthermore, we give the gtk_combo_box_text_append function a NULL id for
   * the parameter as it is not needed.
   */
  combo_box = gtk_combo_box_text_new ();
  gtk_combo_box_text_append (GTK_COMBO_BOX_TEXT (combo_box), NULL, "Select distribution");
  gtk_combo_box_text_append (GTK_COMBO_BOX_TEXT (combo_box), NULL, "Fedora");
  gtk_combo_box_text_append (GTK_COMBO_BOX_TEXT (combo_box), NULL, "Mint");
  gtk_combo_box_text_append (GTK_COMBO_BOX_TEXT (combo_box), NULL, "Suse");

  /* Choose to set the first row as the active one by default, from the beginning */
  gtk_combo_box_set_active (GTK_COMBO_BOX (combo_box), 0);

  /* Connect the signal emitted when a row is selected to the appropriate
   * callback function.
   */
  g_signal_connect (combo_box,
                    "changed",
                    G_CALLBACK (on_changed),
                    NULL);

  /* Add it to the window */
  gtk_container_add (GTK_CONTAINER (window), combo_box);

  gtk_widget_show_all (window);
}



int
main (int argc, char **argv)
{
  GtkApplication *app;
  int status;

  app = gtk_application_new ("org.gtk.example", G_APPLICATION_FLAGS_NONE);
  g_signal_connect (app, "activate", G_CALLBACK (activate), NULL);
  status = g_application_run (G_APPLICATION (app), argc, argv);
  g_object_unref (app);

  return status;
}
